File: SSAFY_Web_Java_Algorithm/알고리즘(C++, JAVA)/SWEA/Solution_D4_5607_조합_이선수.cpp
//#include <iostream>
//#include <string>
//
//using namespace std;
//
//int N, R;
//int modVal = 1234567891;
//long long ft[1000001];
//long long factorial();
//long long Fermat(long long par, int a);
//int main()
//{
//	int T;
//	cin >> T;
//	ft[1] = 1;
//	for (int t = 1; t <= T; t++)
//	{
//		cin >> N >> R;
//
//		long long son = factorial();
//		long long par = (ft[R] * ft[N - R]) % modVal;
//		long long npar = Fermat(par, modVal - 2);
//		cout << "#" << t << " " << (son*npar) % modVal << endl;
//	}
//}
//
//long long factorial()
//{
//	for (int i = 2; i <= N; i++)
//	{
//		ft[i] = (ft[i - 1] * i) % modVal;
//	}
//	return ft[N];
//}
//
//long long Fermat(long long par, int a)
//{
//	if (a == 0) return 1;
//	long long div = Fermat(par, a / 2);
//	long long mul = (div*div) % modVal;
//	if (a % 2 == 0) 
//		return mul;
//	else 
//		return (mul * par) % modVal;
//
//}

File: SSAFY_Web_Java_Algorithm/알고리즘(C++, JAVA)/SWEA/Solution_D4_5607_조합_이선수.java
package ct01hw_서울_7반_이선수;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Solution_D4_5607_조합_이선수 {
	static int N, R;
	static int modVal = 1234567891;
	static long ft[] = new long[1000001];
	
	public static void main(String[] args) throws NumberFormatException, IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		int T = Integer.parseInt(br.readLine());
		
		for(int t = 1; t <= T; t++)
		{
			st = new StringTokenizer(br.readLine());
			N = Integer.parseInt(st.nextToken());
			R = Integer.parseInt(st.nextToken());
			
			long son = factorial();
			long par = (ft[R]*ft[N-R]) % modVal;
			long npar = Fermat(par, modVal - 2);
			System.out.println("#"+t+" " + (son*npar)%modVal);
		}
	}
	
	public static long factorial() {
		ft[1] = 1;
		for (int i = 2; i <= N; i++)
		{
			ft[i] = (ft[i - 1] * i) % modVal;
		}
		return ft[N];
	}
	public static long Fermat(long par, int a) {
		if (a == 0) return 1;
		long div = Fermat(par, a / 2);
		long mul = (div*div) % modVal;
		if (a % 2 == 0) 
			return mul;
		else 
			return (mul * par) % modVal;
	}
}
